### PR TITLE
add dockerfiles

### PR DESCRIPTION
**Pull request by Yolnan**

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+FROM ghcr.io/ros-industrial-consortium/scan_n_plan_workshop:${TAG}
+
+SHELL ["/bin/bash", "-c"]
+
+ENV DEBIAN_FRONTEND noninteractivej
+
+USER root
+
+# Install Python dependencies
+RUN python3 -m pip install open3d pyquaternion --upgrade
```
Comment (marip8) [inline docker/Dockerfile]: I think this is done now in the SNP image, so I don't think we need this here

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# Run apt update and install useful utilities like sudo
+RUN apt update -y -qq && apt install -y -qq sudo
+
+# Build the repository
+# Bind mount the source directory so as not to unnecessarily copy source code into the docker image
+ARG WORKSPACE_DIR=/opt/snp_blending
+RUN --mount=type=bind,target=${WORKSPACE_DIR}/src/snp_blending \
+  source /opt/snp/install/setup.bash \
+  && cd ${WORKSPACE_DIR} \ 
+  && colcon build --cmake-args \
```
Comment (marip8) [inline docker/Dockerfile]: Still need to run `rosdep` to install `exec_depend`s from the `package.xml`

```suggestion
  && cd ${WORKSPACE_DIR} \ 
 && rosdep install \
    --from-paths src \
    -iry \
  && colcon build --cmake-args \
```

File: docker/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# SNP Blending Docker
+
+## Create the Docker image
+### Download
+Download a pre-built Docker image from the container registry:
+
+```
+docker login ghcr.io
+docker pull ghcr.io/ros-industrial-consortium/scan_n_plan_workshop:<tag>
```
Comment (marip8) [inline docker/README.md:9]:
```suggestion
docker pull ghcr.io/ros-industrial-consortium/snp_blending:<tag>
```

File: docker/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    build:
+      context: ..
+      dockerfile: docker/Dockerfile
+      args:
+        - TAG=humble-master
+    environment:
+      DISPLAY: $DISPLAY
+      XAUTHORITY: $XAUTHORITY
+      NVIDIA_DRIVER_CAPABILITIES: all
+      ROS_LOG_DIR: /tmp
```
Comment (marip8) [inline docker/docker-compose.yml:13]:
```suggestion
      ROS_LOG_DIR: /tmp
      ROS_LOCAL_HOST_ONLY: 1
```

File: docker/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    stdin_open: true
+    tty: true
+    network_mode: host
+    privileged: true
+    user: ${CURRENT_UID}  # CURRENT_UID=$(id -u):$(id -g)
+    volumes:
+      - /tmp/.X11-unix:/tmp/.X11-unix
+      - /etc/hosts:/etc/hosts
+      - /etc/passwd:/etc/passwd:ro
+      - /etc/group:/etc/group:ro
```
Comment (marip8) [inline docker/docker-compose.yml:29]: Mount `/dev` for USB devices

```suggestion
      - /etc/group:/etc/group:ro
      - /etc/shadow:/etc/shadow:ro
      - /dev:/dev
```

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    deploy:
+      resources:
+        reservations:
+          devices:
+            - driver: nvidia
+              count: 1
+              capabilities: [gpu]
```
Comment (marip8) [inline docker/docker-compose.yml]: Add micro-ROS agent for MotoROS2 driver

```suggestion
              capabilities: [gpu]
  # Driver
  micro-ros-agent:
    environment:
      ROS_LOG_DIR: /tmp
      ROS_LOCAL_HOST_ONLY: 1
    container_name: micro-ros-agent
    image: microros/micro-ros-agent:humble
    command: udp4 --port 8888
    stdin_open: true
    tty: true
    privileged: true
    network_mode: host
```

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
+name: Docker
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    paths:
```
Comment (marip8) [inline .github/workflows/docker.yml]:
```suggestion
```

**Review by marip8 — CHANGES_REQUESTED** [review]

File: docker/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      XAUTHORITY: $XAUTHORITY
+      NVIDIA_DRIVER_CAPABILITIES: all
+      ROS_LOG_DIR: /tmp
+    container_name: snp_blending
+    image: ghcr.io/ros-industrial-consortium/snp_blending:humble-master
+    stdin_open: true
+    tty: true
+    network_mode: host
+    privileged: true
+    user: ${CURRENT_UID}  # CURRENT_UID=$(id -u):$(id -g)
```
Comment (marip8) [inline docker/docker-compose.yml:22]: Add the user to the `sudo` group inside the container

```suggestion
    user: ${CURRENT_UID}  # CURRENT_UID=$(id -u):$(id -g)
    group_add:
      - sudo
```

File: docker/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      dockerfile: docker/Dockerfile
+      args:
+        - TAG=humble-master
+    environment:
+      DISPLAY: $DISPLAY
+      XAUTHORITY: $XAUTHORITY
+      # NVIDIA_DRIVER_CAPABILITIES: all
+      ROS_LOG_DIR: /tmp
+      ROS_LOCALHOST_ONLY: 1
+      ROS_DOMAIN_ID: 18
```
Comment (marip8) [inline docker/docker-compose.yml]:
```suggestion
      ROS_LOCALHOST_ONLY: $ROS_LOCALHOST_ONLY
      ROS_DOMAIN_ID: $ROS_DOMAIN_ID
```

**Comment by Yolnan** [comment]
@marip8 I'll need to fix start.launch.xml since the realsense node isn't loading the color resolution and fps parameters.

File: docker/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    #       devices:
+    #         - driver: nvidia
+    #           count: 1
+    #           capabilities: [gpu]
+  # Driver
+  micro-ros-agent:
+    environment:
+      ROS_LOG_DIR: /tmp
+      ROS_LOCALHOST_ONLY: 1
+      ROS_DOMAIN_ID: 18
```
Comment (marip8) [inline docker/docker-compose.yml]:
```suggestion
      ROS_LOCALHOST_ONLY: $ROS_LOCALHOST_ONLY
      ROS_DOMAIN_ID: $ROS_DOMAIN_ID
```

File: docker/docker-compose.yml (hunk 9 above)
Comment (Yolnan) [reply docker/docker-compose.yml:15]: Made the change to lines 43 and 44 as well.

**Review by marip8 — APPROVED** [review]

**Pull request merged**